Failed tests (10801): TestGetTruncatableIndexes deadlock on store.mu

The following test appears to have failed:

[#10801](https://circleci.com/gh/cockroachdb/cockroach/10801):

```
I0108 08:59:27.579113 939 storage/replica.go:1619  gossiping system config from store 1, range 1
I0108 08:59:27.579923 939 storage/replica.go:1568  gossiping sentinel from store 1, range 1
I0108 08:59:27.580263 939 storage/replica.go:1574  gossiping first range from store 1, range 1
I0108 08:59:27.584087 939 storage/raft_log_queue.go:74  the raft group doesn't exist for range 100
I0108 08:59:27.597591 939 storage/raft_log_queue.go:133  truncating the raft log of range 1 to 22
panic: test timed out after 5m0s

goroutine 566 [running]:
testing.startAlarm.func1()
	/tmp/workdir/go/src/testing/testing.go:703 +0x16b
created by time.goFunc
	/tmp/workdir/go/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x1a691a8, 0x21162a0, 0xdb, 0xdb, 0x1)
	/tmp/workdir/go/src/testing/testing.go:562 +0xafa
testing.(*M).Run(0xc820033f10, 0x1)
	/tmp/workdir/go/src/testing/testing.go:494 +0xe5
github.com/cockroachdb/cockroach/util/leaktest.TestMainWithLeakCheck(0xc820033f10)
	/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x2f
github.com/cockroachdb/cockroach/storage_test.TestMain(0xc820033f10)
	/go/src/github.com/cockroachdb/cockroach/storage/main_test.go:36 +0x34
main.main()
	github.com/cockroachdb/cockroach/storage/_test/_testmain.go:498 +0x20a

goroutine 17 [syscall, 4 minutes, locked to thread]:
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 5 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x21714e0)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1029 +0x76
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:610 +0x117

goroutine 558 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da8a0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 552 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*StorePool).start.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store_pool.go:213 +0x208
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da1e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 540 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func3()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:556 +0x15d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc82000f0c0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 560 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da8e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 562 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da920)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 538 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/util.ListenAndServe.func2()
	/go/src/github.com/cockroachdb/cockroach/util/net.go:88 +0x7b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc82000f080)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 539 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc8202d83b8)
	/tmp/workdir/go/src/runtime/sema.go:43 +0x26
sync.(*RWMutex).Lock(0xc8202d83b0)
	/tmp/workdir/go/src/sync/rwmutex.go:87 +0xc7
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1595 +0x17d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc820277600)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 564 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*replicaScanner).waitAndProcess(0xc82016c300, 0xece21716f, 0x22bccd0e, 0x2170f60, 0xc8200156e0, 0xc8201380e0, 0xc8201a09c0, 0xc8202d83c0)
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:176 +0x641
github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1.1(0xc8201a09c0, 0xc8202ea240)
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:217 +0xb0
github.com/cockroachdb/cockroach/storage.(*storeRangeSet).Visit(0xc8202ea3c0, 0xc82026ecc0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:225 +0x45c
github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:219 +0x452
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da960)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 541 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc8202d83bc)
	/tmp/workdir/go/src/runtime/sema.go:43 +0x26
sync.(*RWMutex).RLock(0xc8202d83b0)
	/tmp/workdir/go/src/sync/rwmutex.go:36 +0x76
github.com/cockroachdb/cockroach/storage.(*Store).LookupReplica(0xc8202d8200, 0x2b1f0e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:833 +0x63
github.com/cockroachdb/cockroach/storage.(*Store).maybeGossipFirstRange(0xc8202d8200, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:668 +0x3b9
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:623 +0x2b8
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc82000f100)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 535 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc8201d01e0)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:123 +0x5b
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:124 +0x581

goroutine 559 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da8c0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 542 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc8202d83bc)
	/tmp/workdir/go/src/runtime/sema.go:43 +0x26
sync.(*RWMutex).RLock(0xc8202d83b0)
	/tmp/workdir/go/src/sync/rwmutex.go:36 +0x76
github.com/cockroachdb/cockroach/storage.(*Store).LookupReplica(0xc8202d8200, 0xc8201486a0, 0x9, 0x10, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:833 +0x63
github.com/cockroachdb/cockroach/storage.(*Store).maybeGossipSystemConfig(0xc8202d8200, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:684 +0x85
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func2()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:642 +0x2b8
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc82000f120)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 551 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc8202d83bc)
	/tmp/workdir/go/src/runtime/sema.go:43 +0x26
sync.(*RWMutex).RLock(0xc8202d83b0)
	/tmp/workdir/go/src/sync/rwmutex.go:36 +0x76
github.com/cockroachdb/cockroach/storage.(*Store).RaftStatus(0xc8202d8200, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:866 +0x4f
github.com/cockroachdb/cockroach/storage.getTruncatableIndexes(0xc8201a09c0, 0xc820276390, 0x8cdbc4, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft_log_queue.go:71 +0xbd
github.com/cockroachdb/cockroach/storage.(*raftLogQueue).shouldQueue(0xc820092960, 0x0, 0xc800000030, 0xc8201a09c0, 0xc8202da700, 0xc820279500, 0xc8200156e8)
	/go/src/github.com/cockroachdb/cockroach/storage/raft_log_queue.go:111 +0x39
--
testing.tRunner(0xc820078000, 0x2116618)
	/tmp/workdir/go/src/testing/testing.go:456 +0xdd
created by testing.RunTests
	/tmp/workdir/go/src/testing/testing.go:561 +0xaa4

goroutine 563 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*Replica).addWriteCmd(0xc8201a09c0, 0x7fc7b1fcd9f8, 0xc82026f140, 0x0, 0x2e, 0x100000001, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:926 +0x7be
github.com/cockroachdb/cockroach/storage.(*Replica).Send(0xc8201a09c0, 0x7fc7b1fcd9f8, 0xc82026f140, 0x0, 0x2e, 0x100000001, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:628 +0x6a3
github.com/cockroachdb/cockroach/storage.(*Store).Send(0xc8202d8200, 0x7fc7b1fcd9f8, 0xc82026f140, 0x0, 0x2e, 0x100000001, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1312 +0xd81
github.com/cockroachdb/cockroach/storage.(*testSender).Send(0xc820026000, 0x7fc7b1fcd9b8, 0xc8200137c0, 0x0, 0x0, 0x100000001, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/store_test.go:113 +0x6fd
github.com/cockroachdb/cockroach/client.(*DB).send(0xc820270000, 0xc8201c90e0, 0x1, 0x1, 0x1, 0x61a5a4)
	/go/src/github.com/cockroachdb/cockroach/client/db.go:482 +0x1e5
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da940)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 537 [IO wait, 4 minutes]:
net.runtime_pollWait(0x7fc7b1fcd890, 0x72, 0x7fc7b4064050)
	/tmp/workdir/go/src/runtime/netpoll.go:157 +0x63
net.(*pollDesc).Wait(0xc8201f8ae0, 0x72, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:73 +0x56
net.(*pollDesc).WaitRead(0xc8201f8ae0, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:78 +0x44
net.(*netFD).accept(0xc8201f8a80, 0x0, 0x7fc7b406a148, 0xc8202da260)
	/tmp/workdir/go/src/net/fd_unix.go:408 +0x2f6
net.(*TCPListener).AcceptTCP(0xc820026040, 0x60ce50, 0x0, 0x0)
	/tmp/workdir/go/src/net/tcpsock_posix.go:254 +0x77
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202eb500)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 561 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201380e0, 0xc8202da900)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
FAIL	github.com/cockroachdb/cockroach/storage	300.076s
=== RUN   TestBatchBasics
I0108 08:59:42.867505 997 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I0108 08:59:42.870874 997 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I0108 08:59:42.876840 997 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/storage/store.go
+++ b/storage/store.go
@@ -532,6 +532,7 @@ func (s *Store) Start(stopper *stop.Stopper) error {
 			// and initialize those groups.
 			return false, nil
 		}); err != nil {
+		s.mu.Unlock()
 		return err
 	}
 	s.feed.endScanRanges()
@@ -765,12 +766,20 @@ func (s *Store) DisableRaftLogQueue(disabled bool) {
 // need their raft logs truncated and then process each of them.
 // Exposed only for testing.
 func (s *Store) ForceRaftLogScanAndProcess(t util.Tester) {
-	// Add each range to the queue.
+	// MaybeAdd can transitively call back into Store (namely RaftStatus) and end up
+	// attempting to re-acquire mu.RLock, making a deadlock is possible (if Lock called in-between).
+	// Instead, make a copy and release lock before calling MaybeAdd.
+	replicas := make([]*Replica, 0, len(s.replicas))
 	s.mu.RLock()
 	for _, r := range s.replicas {
-		s.raftLogQueue.MaybeAdd(r, s.ctx.Clock.Now())
+		replicas = append(replicas, r)
 	}
 	s.mu.RUnlock()
+
+	// Add each range to the queue.
+	for _, r := range replicas {
+		s.raftLogQueue.MaybeAdd(r, s.ctx.Clock.Now())
+	}
 
 	s.raftLogQueue.DrainQueue(s.ctx.Clock)
 }

